docs/connections.md

```diff
@@ -0,0 +1,597 @@
+variable names appearing in the pattern context. If the identifier was not
+defined, it will appear in the current scope.
+
+For example:
+```
+5 : «a»?
+```
+
+This introduces constraint on the type `a` that its value set must include atom
+`5`. If the `a` is already visible in the scope, this constraint will be added.
+Otherwise, `a` will be introduced into the current scope with that constraint.
+
+// TODO: What if parent scope only ascribes identifier with type constraint but
+only nested scope assigns to it? 
+
+// TODO: Open design question, if `a` should modify existing variable or should
+always try to shadow it. What is the difference between `5 : a` and `5 : A`?
+(except the latter not being able to ever introduce a new identifier)
+
+// TODO: Is it legal or sensible to ascribe a variable when there is no binding?
+
+
+When type is known, the type ascription can be used to constrain type of the
+value:
+
+```
+»a« : 5
+```
+
+This says that value of `a` is of type `5`. Type `5` has only a single allowed
+value: `5`. This will tell compiler to error out if program tries to bind `5`
+with any value that is not known to be `5`.
+
+However, this example refers to some `a` already being visible in scope and does
+not introduce any identifiers.
+
+The `type` in this expression is pattern context and can be used to constrain
+the type variables. It is legal to assign constraints on an identifier using `:`
+multiple times in any of the scopes where identifier is visible.
+
+Signatures are just type ascriptions that happen to precede the assignments.
+They have no special rules currently defined. This area needs further design
+work.
+
+TODO: Open question: does empty type (`Void`) exists in the language?
+
+TODO signatures and their relation with scoping. Difference for root and
+non-root definitions.
+
+Examples:
+
+```
+add : a -> a -> a
+add a b = a + b
+```
+
+TODO: 
+* With current rules `a` from the signature gets introduced into parent scope
+  will be unified with other uses of `a` in other definitions.
+* or, actually, we want this to happen only in the root scope. When in
+  definition body, the `a` actually should be unified between signatures.
+  Doesn't sound that clean though.
+* Does argument-introduced `a` shadows the signature-introduced `a`?
+
+TODO: `b:A` — does this add constraint to `a` or just `b` ? 
+
+# Current engine limitations
+Note: "current" means "in the scope of the first alpha release of enso",
+not "at the moment of writing this document". 
+
+## Extension methods
+The extensions methods (taking `this` as the first parameter) can be defined
+only using the sugared syntax.
+
+While both 
+```
+foo this:a = print "hello"
+```
+
+and 
+
+```
+a.foo = print "hello"
+```
+
+are equivalent, engine currently supports only the latter.
+
+IDE can assume that all extension methods will be introduced using the
+`Type.name` syntax sugar.
+
+// TODO what if non-first argument is named `this` ? Is the magic happening only
```

Comment:
It is an error to define `this` at any position other than the first in the argument list.